LongestPole/Main.lean

```diff
@@ -0,0 +1,191 @@
+deriving ToJson, FromJson
+
+structure SpeedCenterRun_3 where
+  id : String
+  source : SpeedCenterSource_2
+  result : SpeedCenterResult
+deriving ToJson, FromJson
+
+structure SpeedCenterRunResponse where
+  run : SpeedCenterRun_3
+deriving ToJson, FromJson
+
+structure SpeedCenterMessage where
+  repo_id : String
+  message : String
+  commit_hash : String
+deriving ToJson, FromJson
+
+def SpeedCenterRunResponse.instructions (response : SpeedCenterRunResponse) :
+    NameMap Float := Id.run do
+  let mut r : NameMap Float := ∅
+  for m in response.run.result.measurements do
+    let n := m.dimension.benchmark
+    if n.startsWith "~" then
+      r := r.insert (n.drop 1).toName m.value
+  return r
+
+def speedCenterRunResponse (run : String) : IO SpeedCenterRunResponse := do
+  let r ← speedCenterRunJson run
+  match Json.parse r with
+  | .error e => throw <| IO.userError s!"Could not parse speed center JSON: {e}\n{r}"
+  | .ok j => match fromJson? j with
+    | .ok v => pure v
+    | .error e => match fromJson? j with
+      | .ok (v : SpeedCenterMessage) =>
+        IO.eprintln s!"http://speed.lean-fro.org says: {v.message}"
+        IO.eprintln s!"Try moving to an older commit?"
+        IO.Process.exit 1
+      | .error _ => throw <| IO.userError s!"Could not parse speed center JSON: {e}\n{j}"
+
+def headSha : IO String := return (← runCmd "git" #["rev-parse", "HEAD"]).trim
+
+def speedCenterInstructions (run : String) : IO (NameMap Float) :=
+  return (← speedCenterRunResponse run).instructions
+
+partial def cumulativeInstructions (instructions : NameMap Float) (graph : NameMap (Array Name)) :
+    NameMap Float :=
+  graph.fold (init := ∅) fun m n _ => go n m
+where
+  go (n : Name) (m : NameMap Float) : NameMap Float :=
+    if m.contains n then
+      m
+    else
+      let parents := (graph.find? n).getD #[]
+      let m := parents.foldr (init := m) fun n' m => go n' m
+      let t := (parents.map fun n' => (m.find? n').getD 0).foldr max 0
+      m.insert n ((instructions.find? n).getD 0 + t)
+
+def slowestParents (cumulative : NameMap Float) (graph : NameMap (Array Name)) :
+    NameMap Name :=
+  graph.fold (init := ∅) fun m n parents =>
+    match parents.toList with
+    | [] => m
+    | h :: t => Id.run do
+      let mut r := h
+      for n' in t do
+        if (cumulative.find? n').getD 0 > (cumulative.find? r).getD 0 then
+          r := n'
+      return m.insert n r
+
+def totalInstructions (instructions : NameMap Float) (graph : NameMap (Array Name)) :
+    NameMap Float :=
+  let transitive := graph.transitiveClosure
+  transitive.filterMap fun n s => some <| s.fold (init := (instructions.find? n).getD 0)
+    fun t n' => t + ((instructions.find? n').getD 0)
+
+def String.rightPad (s : String) (n : Nat) (c : Char := ' ') : String :=
+  s ++ String.replicate (n - s.length) c
+
+def String.leftPad (s : String) (n : Nat) (c : Char := ' ') : String :=
+  String.replicate (n - s.length) c ++ s
```

Comment:
Are these not yet available in some lower level lib? I'm fine with having them here for now...
What is the risk that others will try to import this file and get a name conflict?

Reply:
Risk of this being imported seems minimal, since the point of this file is just to be the `Main` file for the executable. I'll put a todo note to move them.

Reply:
Hmm, perhaps we should have been more worried about a _lack_ of name conflict: I am now observing that a `String.leftpad` already exists. Unfortunately `rightpad` does not.

Reply:
Ok I have now

* deleted `leftPad` in favor of the preexisting `leftpad`
* kept the existing `leftpad` type signature, but added the default argument whitespace.
* moved/renamed `rightPad` to match.